### DISCUSSION TITLE
[Idea] Add a hook in the PaymentModule::ValidateOrder() method to customize the order confirmation email

### DISCUSSION THREAD

**Discussion by Codencode** [PrestaShop/PrestaShop#36913] [Ideas]
### The problem

In some cases, there is a need to send an order confirmation email that is different from the default one. For a client, for instance, it is sometimes necessary to send an order confirmation email without addresses, carrier, and shipping costs.

### The solution

The solution could be the addition of a new hook in the PaymentModule::ValidateOrder() method, right before sending the order confirmation email to the user.

https://github.com/PrestaShop/PrestaShop/blob/c6cf71fb6e8363b0b25fcbc450e25f7ae4cb0ced/classes/PaymentModule.php#L699-L722
 This way, if needed, it will be possible to manage the sending of different order confirmation emails based on the type of product, obviously through the development of a custom module.
 
 ---
 A possible solution could be:
 
 ```php
                     $sendEmailResult = Hook::exec(
                        'actionSendEmailOrderConf',
                        [
                            'data' => $data,
                            'file_attachement' => $file_attachement,
                            'order' => $order,
                        ],
                        null,
                        true
                    );

                    if ($sendEmailResult === null) {
                        $sendEmail = false;
                    } else {
                        $sendEmail = array_reduce(
                            $sendEmailResult,
                            function ($carry, $item) {
                                return ($item === false) ? false : $carry;
                            },
                            true
                        );
                    }

                    if ($sendEmail) {
                        Mail::Send(
                            (int) $order->id_lang,
                            'order_conf',
                            $this->context->getTranslator()->trans(
                                'Order confirmation',
                                [],
                                'Emails.Subject',
                                $orderLanguage->locale
                            ),
                            $data,
                            $this->context->customer->email,
                            $this->context->customer->firstname . ' ' . $this->context->customer->lastname,
                            null,
                            null,
                            $file_attachement,
                            null,
                            _PS_MAIL_DIR_,
                            false,
                            (int) $order->id_shop
                        );
                    }
```

Obviously, a better name for the hook would be needed.

This solution is similar to the one present in the Mail::send() method.

https://github.com/PrestaShop/PrestaShop/blob/c6cf71fb6e8363b0b25fcbc450e25f7ae4cb0ced/classes/Mail.php#L154-L190

### Do you plan to work on this subject?

- [X] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [X] I'm willing to provide any wireframes or design assets required for this feature.
- [X] I'm willing to submit a Pull Request that implements this feature.
- [x] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

**Comment by Hlavtox** [comment]
I think this need can be universally described as "Stop the mail send process".

You could use it to send your own custom email.
I for example needed it to NOT send some emails in some cases, which I did not found a way to do.

**Reply by Codencode** [reply]
@Hlavtox,
Are you referring to the title of the issue? Should I change the title?

**Reply by Hlavtox** [reply]
@Codencode Nono, just the general idea of what you need and how to implement it in an universal way. :-)

We could add a hook in the mail sending process with an option to "intercept" and stop the mail sending process. Than we could do whatever we want.

**Reply by Codencode** [reply]
A ok,

actually the hook `actionEmailSendBefore` already exists

https://github.com/PrestaShop/PrestaShop/blob/c6cf71fb6e8363b0b25fcbc450e25f7ae4cb0ced/classes/Mail.php#L155-L193

But I would like something specific to change the order confirmation email.

For my client I used the hook `actionEmailSendBefore` to be able to differentiate, but I think a specific hook could be better

**Reply by Hlavtox** [reply]
Ah, I did not know this. :-)

Well, the more hooks you create, the more database queries and more calls it creates. If there is a hook that is easy to use, and this one is, I think it's OK no?

Or you can just use a hook to change the email template and add some more variables, done.

**Reply by Codencode** [reply]
Yes, of course,

but the `actionEmailSendBefore` hook is executed for every email sent, so to change the order confirmation email you need to run the check for every email sent.
Instead, if you create a new hook and run it in the `PaymentModule::ValidateOrder()` method, it would only be executed in case of an order.
